src/components/organisms/AppRouter/AdminSubrouter.tsx

```diff
@@ -10,26 +10,31 @@ import VenueWizard_v2 from "pages/Admin/Venue/VenueWizard/VenueWizard";
 import { AdminAdvancedSettings } from "pages/AdminAdvancedSettings";
 
 import { AdminVenueView } from "components/organisms/AdminVenueView";
+import { WorldUsersProvider } from "hooks/users";
+import { useVenueId } from "hooks/useVenueId";
 
 export const AdminSubrouter: React.FC = () => {
+  const venueId = useVenueId();
   return (
-    <Switch>
-      {/* Admin V1 */}
-      <Route path="/admin/venue/rooms/:venueId" component={RoomsForm} />
-      <Route path="/admin/venue/creation" component={VenueWizard} />
-      <Route path="/admin/venue/edit/:venueId" component={VenueWizard} />
-      <Route path="/admin/:venueId" component={Admin} />
-      <Route path="/admin" component={Admin} />
+    <WorldUsersProvider venueId={venueId}>
```

Comment:
This is how I was thinking of doing it.. but what I wasn't necessarily sure about was what `venueId` would actually end up being in this instance, since we're reading it from the 'routers path', yet that isn't defined yet in theory, since it happens in the `Route`s below.

Can you do some hacky console logs or similar and tell me what it returns when eg. on `/admin-ng` vs when on `/admin-ng/edit/:venueId` for example.

Reply:
The not having `:venueId` is the reason why I wanted to try wrapping only the components with those routes (and proved more work for less benefit),
but in general, we can just let the provider do the job only on existing id.

Will test.

Reply:
Provider already has a safeguard for missing `venueId` among other things:
```
  const shouldConnect =
    !isSovereignVenueLoading &&
    isDefined(sovereignVenueId) &&
    isDefined(venueId);

  // TODO: refactor this to not use useFirestoreConnect
  useFirestoreConnect(() => {
    if (!shouldConnect) return [];
```

Can't find edge case where missing id would display console errors.
Testing the `/admin-ng` and `admin` without it in parameters correctly displays message that a venue should be selected or just recovers it from the subsequent navigation URL which contains it.

Reply:
This is what I was requesting above:

Using this hacky console log:
```diff
diff --git a/src/components/organisms/AppRouter/AdminSubrouter.tsx b/src/components/organisms/AppRouter/AdminSubrouter.tsx
index 30d7793b4..e7913b567 100644
--- a/src/components/organisms/AppRouter/AdminSubrouter.tsx
+++ b/src/components/organisms/AppRouter/AdminSubrouter.tsx
@@ -16,6 +16,9 @@ import { AdminVenueView } from "components/organisms/AdminVenueView";
 
 export const AdminSubrouter: React.FC = () => {
   const venueId = useVenueId();
+
+  console.log(venueId, window.location);
+
   return (
     <WorldUsersProvider venueId={venueId}>
       <Switch>

```

This is when accessing `/admin-ng`:

![image](https://user-images.githubusercontent.com/753891/125064735-505f5880-e0f4-11eb-9fda-72ba7b4631af.png)

This is when accessing `/admin-ng/venue/devaliashacksville`:
![image](https://user-images.githubusercontent.com/753891/125064790-5c4b1a80-e0f4-11eb-96f1-e052e3c0bffa.png)

Note that in both cases the `venueId` is `undefined`, which is invalid. So this isn't going to work for us.

Reply:
This will solve it though:

```diff
diff --git a/src/components/organisms/AppRouter/AdminSubrouter.tsx b/src/components/organisms/AppRouter/AdminSubrouter.tsx
index 30d7793b4..18e60d7e2 100644
--- a/src/components/organisms/AppRouter/AdminSubrouter.tsx
+++ b/src/components/organisms/AppRouter/AdminSubrouter.tsx
@@ -15,27 +15,69 @@ import { AdminAdvancedSettings } from "pages/AdminAdvancedSettings";
 import { AdminVenueView } from "components/organisms/AdminVenueView";
 
 export const AdminSubrouter: React.FC = () => {
-  const venueId = useVenueId();
   return (
-    <WorldUsersProvider venueId={venueId}>
-      <Switch>
-        {/* Admin V1 */}
-        <Route path="/admin/venue/rooms/:venueId" component={RoomsForm} />
-        <Route path="/admin/venue/creation" component={VenueWizard} />
-        <Route path="/admin/venue/edit/:venueId" component={VenueWizard} />
-        <Route path="/admin/:venueId" component={Admin} />
-        <Route path="/admin" component={Admin} />
-
-        {/* Admin V2/3/NG */}
-        <Route path="/admin-ng/venue/:venueId?" component={AdminVenueView} />
-        <Route
-          path="/admin-ng/advanced-settings/:venueId?"
-          component={AdminAdvancedSettings}
-        />
-        <Route path="/admin-ng/venue/creation" component={VenueWizard_v2} />
-        <Route path="/admin-ng/edit/:venueId" component={VenueWizard_v2} />
-        <Route path="/admin-ng" component={Admin_v2} />
-      </Switch>
-    </WorldUsersProvider>
+    <Switch>
+      {/* Admin V1 */}
+      <Route path="/admin/venue/rooms/:venueId">
+        <WithWorldUsersProviderForVenueId>
+          <RoomsForm />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+      <Route path="/admin/venue/creation">
+        <WithWorldUsersProviderForVenueId>
+          <VenueWizard />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+      <Route path="/admin/venue/edit/:venueId">
+        <WithWorldUsersProviderForVenueId>
+          <VenueWizard />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+      <Route path="/admin/:venueId">
+        <WithWorldUsersProviderForVenueId>
+          <Admin />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+      <Route path="/admin">
+        <WithWorldUsersProviderForVenueId>
+          <Admin />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+
+      {/* Admin V2/3/NG */}
+      <Route path="/admin-ng/venue/:venueId?">
+        <WithWorldUsersProviderForVenueId>
+          <AdminVenueView />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+      <Route path="/admin-ng/advanced-settings/:venueId?">
+        <WithWorldUsersProviderForVenueId>
+          <AdminAdvancedSettings />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+      <Route path="/admin-ng/venue/creation">
+        <WithWorldUsersProviderForVenueId>
+          <VenueWizard_v2 />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+      <Route path="/admin-ng/edit/:venueId">
+        <WithWorldUsersProviderForVenueId>
+          <VenueWizard_v2 />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+      <Route path="/admin-ng">
+        <WithWorldUsersProviderForVenueId>
+          <Admin_v2 />
+        </WithWorldUsersProviderForVenueId>
+      </Route>
+    </Switch>
   );
 };
+
+const WithWorldUsersProviderForVenueId: React.FC = ({ children }) => {
+  const venueId = useVenueId();
+
+  console.log(venueId, window.location);
+
+  return <WorldUsersProvider venueId={venueId}>{children}</WorldUsersProvider>;
+};
```

Reply:
Something like this could also potentially work:

```typescript
const RouteWithWorldUsersProvider: React.FC<RouteWithWorldUsersProviderProps> = ({
  path,
  children,
}) => {
  const venueId = useVenueId();

  console.log(venueId, window.location);

  return (
    <Route {...{ path }}>
      <WorldUsersProvider venueId={venueId}>{children}</WorldUsersProvider>
    </Route>
  );
};
```

Then used like this:
```
{/*<Route path="/admin-ng" component={Admin_v2} />*/}
<RouteWithWorldUsersProvider path="/admin-ng">
  <Admin_v2 />
</RouteWithWorldUsersProvider>
```

Reply:
K, maybe it looks less flexible, but it does provide us with opportunities to:

- keep it DRY in case of boilerplate for all the routes
- allow us to include logic that may add one or more providers based on props (boolean flags I guess)

Reply:
And clean up the component at https://github.com/sparkletown/sparkle/pull/1769#discussion_r666860767

Reply:
https://github.com/sparkletown/sparkle/pull/1769#discussion_r666854028 Would seem the better choice to me

Reply:
Added wrapper component named `Provided`  https://github.com/sparkletown/sparkle/pull/1769/commits/5efe4611ba289afe281b83ccf1106504ed793b68

Reply:
Had to add exception for one route (at aa35a4db8b00596596a6fb5b9e8edf3c0a1839ab) , didn't delve deeper into the cause

```
<Route
        path="/admin/venue/creation"
        component={
          // for some state related reasons, Provided fails this VenueWizard
          // and using function here will cause the error
          // so WithNavigationBar wraps the RelatedVenuesProvider for it
          VenueWizard
        }
      />
```

Reply:
I'm pretty sure that's going to have the exact same issues as pointed out in https://github.com/sparkletown/sparkle/pull/1769#discussion_r666850256 since we need the path from `Route` to be able to know how to extract the `:venueId` from it. Therefore we can't use any form of 'parent wrapper' to do it.

Reply:
As implemented in https://github.com/sparkletown/sparkle/commit/5efe4611ba289afe281b83ccf1106504ed793b68, i'm pretty sure that's going to have the exact same issues as pointed out in https://github.com/sparkletown/sparkle/pull/1769#discussion_r666850256 since we need the path from `Route` to be able to know how to extract the `:venueId` from it. Therefore we can't use any form of 'parent wrapper' to do it.

Reply:
In general, having two providers, one higher and one lower in the component tree, will there be issues of hooks getting the data from the closest ancestor (the lower level provider) or some other errors?

Reply:
The 'lowest' provider will 'win' and will result in duplicated data/etc if stored in the provider itself I believe.

Reply:
I've looked at the code that exists now and it seems to work fine. Resolving this.